### PR TITLE
Crash fixes

### PR DESCRIPTION
**Pull request by flodiebold**
This fixes a bunch of crashes found while running type inference on the whole rustc repo :sweat_smile: 
 - avoid infinite recursion with ref bind patterns
 - avoid another infinite recursion
 - handle literal patterns, add a new LITERAL_PAT syntax node for this
 - fix an expect that's wrong on some invalid code

### CODE REVIEW

File: crates/ra_hir/src/ty.rs [hunk 1]
```diff
@@ -989,19 +1000,24 @@ impl<'a, D: HirDatabase> InferenceContext<'a, D> {
...
     fn resolve_ty_shallow<'b>(&mut self, ty: &'b Ty) -> Cow<'b, Ty> {
-        match ty {
-            Ty::Infer(tv) => {
-                let inner = tv.to_inner();
-                match self.var_unification_table.probe_value(inner).known() {
-                    Some(known_ty) => {
-                        // The known_ty can't be a type var itself
-                        Cow::Owned(known_ty.clone())
+        let mut ty = Cow::Borrowed(ty);
+        for _ in 0..3 {
```
Comment (kjeremy) [inline crates/ra_hir/src/ty.rs]: This might deserve a comment (I'm assuming it's arbitrary). Maybe a message if we fail to resolve in that time?

**Comment by kjeremy** [comment]
Nice! How did you run inference over the repo? That's a great test.

File: crates/ra_hir/src/ty.rs (hunk 1 above)
Comment (matklad) [reply crates/ra_hir/src/ty.rs]: A mark might be useful here as well!

**Comment by kjeremy** [comment]
Does this fix #717 ?

**Comment by matklad** [comment]
@kjeremy unlikely, #717 is about "unknown files", so, something VFS related.

**Comment by flodiebold** [comment]
> Nice! How did you run inference over the repo? That's a great test.
See #774 ;)

**Comment by flodiebold** [comment]
> Does this fix #717 ?

No, all crashes I found were in type inference.

File: crates/ra_hir/src/ty.rs (hunk 1 above)
Comment (flodiebold) [reply crates/ra_hir/src/ty.rs]: It's not quite arbitrary -- we have normal, int and float type variables, and they might be nested, but each only once (even an int variable resolving to a float variable shouldn't happen, but better to make sure).

I added a mark and comment.

**Pull request merged**